In [1]:
import pandas as pd

In [2]:
#Load Data
customers = pd.read_csv('../data/raw/customers.csv')
offers = pd.read_csv('../data/raw/offers.csv')
events = pd.read_csv('../data/raw/events.csv')
data_dic = pd.read_csv('../data/raw/data_dictionary.csv')


In [3]:
customers.head()

,customer_id,became_member_on,gender,age,income
0,68be06ca386d4c31939f3a4f0e3dd783,20170212,NaN,118,NaN
1,0610b486422d4921ae7d2bf64640c50b,20170715,F,55,112000.0
2,38fe809add3b4fcf9315a9694bb96ff5,20180712,NaN,118,NaN
3,78afa995795e4d85b5d9ceeca43f5fef,20170509,F,75,100000.0
4,a03223e636434f42ac4c3df47e8bac43,20170804,NaN,118,NaN


In [4]:

customers.info()
customers.describe()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17000 entries, 0 to 16999
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customer_id       17000 non-null  object 
 1   became_member_on  17000 non-null  int64  
 2   gender            14825 non-null  object 
 3   age               17000 non-null  int64  
 4   income            14825 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 664.2+ KB


,became_member_on,age,income
count,1.700000e+04,17000.000000,14825.000000
mean,2.016703e+07,62.531412,65404.991568
std,1.167750e+04,26.738580,21598.299410
min,2.013073e+07,18.000000,30000.000000
25%,2.016053e+07,45.000000,49000.000000
50%,2.017080e+07,58.000000,64000.000000
75%,2.017123e+07,73.000000,80000.000000
max,2.018073e+07,118.000000,120000.000000


In [5]:
offers.head()

,offer_id,offer_type,difficulty,reward,duration,channels
0,ae264e3637204a6fb9bb56bc8210ddfd,bogo,10,10,7,"['email', 'mobile', 'social']"
1,4d5c57ea9a6940dd891ad53e9dbe8da0,bogo,10,10,5,"['web', 'email', 'mobile', 'social']"
2,3f207df678b143eea3cee63160fa8bed,informational,0,0,4,"['web', 'email', 'mobile']"
3,9b98b8c7a33c4b65b9aebfe6a799e6d9,bogo,5,5,7,"['web', 'email', 'mobile']"
4,0b1e1539f2cc45b7b9fa7c272da2e1d7,discount,20,5,10,"['web', 'email']"


In [6]:

offers.info()
offers.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   offer_id    10 non-null     object
 1   offer_type  10 non-null     object
 2   difficulty  10 non-null     int64 
 3   reward      10 non-null     int64 
 4   duration    10 non-null     int64 
 5   channels    10 non-null     object
dtypes: int64(3), object(3)
memory usage: 612.0+ bytes


,difficulty,reward,duration
count,10.000000,10.000000,10.000000
mean,7.700000,4.200000,6.500000
std,5.831905,3.583915,2.321398
min,0.000000,0.000000,3.000000
25%,5.000000,2.000000,5.000000
50%,8.500000,4.000000,7.000000
75%,10.000000,5.000000,7.000000
max,20.000000,10.000000,10.000000


In [7]:
events.head()

,customer_id,event,value,time
0,78afa995795e4d85b5d9ceeca43f5fef,offer received,{'offer id': '9b98b8c7a33c4b65b9aebfe6a799e6d9'},0
1,a03223e636434f42ac4c3df47e8bac43,offer received,{'offer id': '0b1e1539f2cc45b7b9fa7c272da2e1d7'},0
2,e2127556f4f64592b11af22de27a7932,offer received,{'offer id': '2906b810c7d4411798c6938adc9daaa5'},0
3,8ec6ce2a7e7949b1bf142def7d0e0586,offer received,{'offer id': 'fafdcd668e3743c1bb461111dcafc2a4'},0
4,68617ca6246f4fbc85e91a2a49552598,offer received,{'offer id': '4d5c57ea9a6940dd891ad53e9dbe8da0'},0


In [8]:
events.info()
events.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 306534 entries, 0 to 306533
Data columns (total 4 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   customer_id  306534 non-null  object
 1   event        306534 non-null  object
 2   value        306534 non-null  object
 3   time         306534 non-null  int64 
dtypes: int64(1), object(3)
memory usage: 9.4+ MB


,time
count,306534.000000
mean,366.382940
std,200.326314
min,0.000000
25%,186.000000
50%,408.000000
75%,528.000000
max,714.000000


# 1. Data Cleaning
## 1.1 Clean Customer Table

In customer table, we could see that when gender and income are null then age is 118. This likely represents missing or unknown customer data, not real age. So we replace 118 to NA.

In [9]:
customers['age'] = customers['age'].replace(118, pd.NA)


We also handle missing value of gender and income

In [10]:
customers.isnull().sum()

customer_id            0
became_member_on       0
gender              2175
age                 2175
income              2175
dtype: int64

In [11]:
customers['gender'] = customers['gender'].fillna('Unknown')

In [12]:
customers['income'] = customers['income'].fillna(customers['income'].median())

We also convert data type of became_member_on to date

In [13]:
customers['became_member_on'] = pd.to_datetime(customers['became_member_on'], format='%Y%m%d')
customers.head()

,customer_id,became_member_on,gender,age,income
0,68be06ca386d4c31939f3a4f0e3dd783,2017-02-12,Unknown,<NA>,64000.0
1,0610b486422d4921ae7d2bf64640c50b,2017-07-15,F,55,112000.0
2,38fe809add3b4fcf9315a9694bb96ff5,2018-07-12,Unknown,<NA>,64000.0
3,78afa995795e4d85b5d9ceeca43f5fef,2017-05-09,F,75,100000.0
4,a03223e636434f42ac4c3df47e8bac43,2017-08-04,Unknown,<NA>,64000.0


## 1.2 Clean Events Table
The events dataset contains customer activities such as receiving, viewing, and completing offers, as well as transactions. However, the value column is semi-structured and stores different information depending on the event type. Offer-related events contain an offer id, while transaction events contain an amount.

In [14]:
events['value'].head()

0    {'offer id': '9b98b8c7a33c4b65b9aebfe6a799e6d9'}
1    {'offer id': '0b1e1539f2cc45b7b9fa7c272da2e1d7'}
2    {'offer id': '2906b810c7d4411798c6938adc9daaa5'}
3    {'offer id': 'fafdcd668e3743c1bb461111dcafc2a4'}
4    {'offer id': '4d5c57ea9a6940dd891ad53e9dbe8da0'}
Name: value, dtype: object

In [15]:
events['event'].value_counts()

event
transaction        138953
offer received      76277
offer viewed        57725
offer completed     33579
Name: count, dtype: int64

In [16]:
events[events['event'] == 'transaction']['value'].head()

12654    {'amount': 0.8300000000000001}
12657                 {'amount': 34.56}
12659                 {'amount': 13.23}
12670                 {'amount': 19.51}
12671                 {'amount': 18.97}
Name: value, dtype: object

In [17]:
# Convert value column from string to dictionary

import ast
events['value'] = events['value'].apply(ast.literal_eval)



In [18]:
# extract relevent fields into seperate columns

events['offer_id'] = events['value'].apply(lambda x: x.get('offer id'))
events['amount'] = events['value'].apply(lambda x: x.get('amount'))

In [19]:
# drop the origianl value column
events.drop('value', axis =1, inplace = True)



In [20]:
events.head()

,customer_id,event,time,offer_id,amount
0,78afa995795e4d85b5d9ceeca43f5fef,offer received,0,9b98b8c7a33c4b65b9aebfe6a799e6d9,NaN
1,a03223e636434f42ac4c3df47e8bac43,offer received,0,0b1e1539f2cc45b7b9fa7c272da2e1d7,NaN
2,e2127556f4f64592b11af22de27a7932,offer received,0,2906b810c7d4411798c6938adc9daaa5,NaN
3,8ec6ce2a7e7949b1bf142def7d0e0586,offer received,0,fafdcd668e3743c1bb461111dcafc2a4,NaN
4,68617ca6246f4fbc85e91a2a49552598,offer received,0,4d5c57ea9a6940dd891ad53e9dbe8da0,NaN


To make the data usable for analysis, the value column was first converted from a string into a dictionary format. Then, relevant fields were extracted into separate columns (offer_id and amount), allowing each type of information to be clearly represented. This transformation standardizes the dataset, making it possible to link events to offers, analyze customer spending, and track the full customer journey effectively.

## 1.1.2 Check Duplicate values

In [21]:
customers.duplicated().sum()

np.int64(0)

In [22]:
events.duplicated().sum()

np.int64(2962)

In [23]:
offers.duplicated().sum()

np.int64(0)

In [24]:
events = events.drop_duplicates()
events.duplicated().sum()

np.int64(0)

The events dataset contained 2,962 fully duplicated records identified through full-row comparison. Since each row represents a unique customer action at a specific time, identical records are unlikely to be valid. For example, if a customer has two identical entries for a transaction at the same timestamp with the same amount, counting both would incorrectly double the revenue. Therefore, these duplicates were removed to prevent double counting and ensure accurate KPI calculation

## 1.1.3 Convert time into days
Time column is in hours since the start of the experiment. So, we convert it to days as offers have duration in days and it is easier to compare offer duration and completion time.


In [25]:
events['time_days'] = events['time']/24
events.tail()

,customer_id,event,time,offer_id,amount,time_days
306529,b3a1272bc9904337b331bf348c3e8c17,transaction,714,None,1.59,29.75
306530,68213b08d99a4ae1b0dcb72aebd9aa35,transaction,714,None,9.53,29.75
306531,a00058cf10334a308c68e7631c529907,transaction,714,None,3.61,29.75
306532,76ddbd6576844afe811f1a3c0fbb5bec,transaction,714,None,3.53,29.75
306533,c02b10e8752c4d8e9b73f918558531f7,transaction,714,None,4.05,29.75


## 1.2 Clean Offer Table
Channels column has a list stored as text which is not directly usable in analysis. So, we transform the multi-valued channels field into binary indicator columns to enable channel-level performance analysis i.e one-hot encoding.

In [26]:
# Convert string to list
offers['channels'] = offers['channels'].apply(ast.literal_eval)
offers['channels']



0         [email, mobile, social]
1    [web, email, mobile, social]
2            [web, email, mobile]
3            [web, email, mobile]
4                    [web, email]
5    [web, email, mobile, social]
6    [web, email, mobile, social]
7         [email, mobile, social]
8    [web, email, mobile, social]
9            [web, email, mobile]
Name: channels, dtype: object

In [27]:
# create dummy columns
channels_dummies = offers['channels'].explode().str.get_dummies().groupby(level=0).max()
offers = pd.concat([offers, channels_dummies], axis=1)
offers.head()


,offer_id,offer_type,difficulty,reward,duration,channels,email,mobile,social,web
0,ae264e3637204a6fb9bb56bc8210ddfd,bogo,10,10,7,"[email, mobile, social]",1,1,1,0
1,4d5c57ea9a6940dd891ad53e9dbe8da0,bogo,10,10,5,"[web, email, mobile, social]",1,1,1,1
2,3f207df678b143eea3cee63160fa8bed,informational,0,0,4,"[web, email, mobile]",1,1,0,1
3,9b98b8c7a33c4b65b9aebfe6a799e6d9,bogo,5,5,7,"[web, email, mobile]",1,1,0,1
4,0b1e1539f2cc45b7b9fa7c272da2e1d7,discount,20,5,10,"[web, email]",1,0,0,1


In [28]:
## Merge data

events_offers = events.merge(offers, how='left', on='offer_id')
full_data = events_offers.merge(customers, how='left', on='customer_id')
full_data.head()





,customer_id,event,time,offer_id,amount,time_days,offer_type,difficulty,reward,duration,channels,email,mobile,social,web,became_member_on,gender,age,income
0,78afa995795e4d85b5d9ceeca43f5fef,offer received,0,9b98b8c7a33c4b65b9aebfe6a799e6d9,NaN,0.0,bogo,5.0,5.0,7.0,"[web, email, mobile]",1.0,1.0,0.0,1.0,2017-05-09,F,75,100000.0
1,a03223e636434f42ac4c3df47e8bac43,offer received,0,0b1e1539f2cc45b7b9fa7c272da2e1d7,NaN,0.0,discount,20.0,5.0,10.0,"[web, email]",1.0,0.0,0.0,1.0,2017-08-04,Unknown,<NA>,64000.0
2,e2127556f4f64592b11af22de27a7932,offer received,0,2906b810c7d4411798c6938adc9daaa5,NaN,0.0,discount,10.0,2.0,7.0,"[web, email, mobile]",1.0,1.0,0.0,1.0,2018-04-26,M,68,70000.0
3,8ec6ce2a7e7949b1bf142def7d0e0586,offer received,0,fafdcd668e3743c1bb461111dcafc2a4,NaN,0.0,discount,10.0,2.0,10.0,"[web, email, mobile, social]",1.0,1.0,1.0,1.0,2017-09-25,Unknown,<NA>,64000.0
4,68617ca6246f4fbc85e91a2a49552598,offer received,0,4d5c57ea9a6940dd891ad53e9dbe8da0,NaN,0.0,bogo,10.0,10.0,5.0,"[web, email, mobile, social]",1.0,1.0,1.0,1.0,2017-10-02,Unknown,<NA>,64000.0


In [29]:
full_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303572 entries, 0 to 303571
Data columns (total 19 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   customer_id       303572 non-null  object        
 1   event             303572 non-null  object        
 2   time              303572 non-null  int64         
 3   offer_id          134002 non-null  object        
 4   amount            138953 non-null  float64       
 5   time_days         303572 non-null  float64       
 6   offer_type        134002 non-null  object        
 7   difficulty        134002 non-null  float64       
 8   reward            134002 non-null  float64       
 9   duration          134002 non-null  float64       
 10  channels          134002 non-null  object        
 11  email             134002 non-null  float64       
 12  mobile            134002 non-null  float64       
 13  social            134002 non-null  float64       
 14  web 

In [30]:
## Save Clean Data
customers.to_csv('../data/cleaned/customers_clean.csv', index=False)
events.to_csv('../data/cleaned/events_clean.csv', index=False)
offers.to_csv('../data/cleaned/offers_clean.csv', index=False)
full_data.to_csv('../data/cleaned/full_data.csv', index=False)

In [31]:
## 2. Connect to Database

In [32]:
#install packages
!pip install sqlalchemy psycopg2-binary pandas

# !pip install python-dotenv
import os
from dotenv import load_dotenv

load_dotenv()  # loads variables from .env


Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 26.0 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


True

In [33]:
# connect jupyter to PostgreSQL
from sqlalchemy import create_engine

# Databse connection details
username = os.getenv("PG_USER")
password = os.getenv("PG_PASSWORD")
host = os.getenv("PG_HOST")
port = os.getenv("PG_PORT")
database = os.getenv("PG_DATABASE")

# Create PostgreSQL engine
engine = create_engine(f"postgresql+psycopg2://{username}:{password}@{host}:{port}/{database}")

# test connection
with engine.connect() as conn:
    print("Connected to PostgreSQL!")

Connected to PostgreSQL!


In [35]:
# load dataframe to Postgres
customers = pd.read_csv("../data/cleaned/customers_clean.csv")
offers = pd.read_csv("../data/cleaned/offers_clean.csv")
events = pd.read_csv("../data/cleaned/events_clean.csv")

# Load into public schema/table (auto-creates table)
customers.to_sql("customers_clean", engine, schema="public", if_exists="replace", index=False)
offers.to_sql("offers_clean", engine, schema="public", if_exists="replace", index=False)
events.to_sql("events_clean", engine, schema="public", if_exists="replace", index=False)

print("Loaded into PostgreSQL: database _ clean datasets")

Loaded into PostgreSQL: database _ clean datasets
